agent/job_runner.go

```diff
@@ -683,20 +683,22 @@ func (r *JobRunner) executePreBootstrapHook(hook string) (bool, error) {
 }
 
 // Starts the job in the Buildkite Agent API. We'll retry on connection-related
-// issues, but if a connection succeeds and we get an error response back from
+// issues, but if a connection succeeds and we get an client error response back from
 // Buildkite, we won't bother retrying. For example, a "no such host" will
-// retry, but a 422 from Buildkite won't.
+// retry, but an HTTP >= 400 <= 500 response from Buildkite won't.
 func (r *JobRunner) startJob(startedAt time.Time) error {
 	r.job.StartedAt = startedAt.UTC().Format(time.RFC3339Nano)
 
 	return roko.NewRetrier(
 		roko.WithMaxAttempts(30),
 		roko.WithStrategy(roko.Constant(5*time.Second)),
 	).Do(func(rtr *roko.Retrier) error {
-		_, err := r.apiClient.StartJob(r.job)
+		response, err := r.apiClient.StartJob(r.job)
 
 		if err != nil {
-			if api.IsRetryableError(err) {
+			if response != nil && api.IsRetryableStatus(response) {
+				r.logger.Warn("%s (%s)", err, rtr)
```

Comment:
I could be wrong, but it looks like there's some response wrapping that happens in `client.doRequest`
```go
	err = checkResponse(resp)
	if err != nil {
		// even though there was an error, we still return the response
		// in case the caller wants to inspect it further
		return response, err
	}
```
so if it's a non-2xx response we can expect err to be non-nil.

Reply:
aaah true! good call, i should've read the code 😅